sampler.go

```diff
@@ -544,16 +554,18 @@ func (s *RemotelyControlledSampler) updateAdaptiveSampler(strategies *sampling.P
 
 // NB: this function should only be called while holding a Write lock
 func (s *RemotelyControlledSampler) updateRateLimitingOrProbabilisticSampler(res *sampling.SamplingStrategyResponse) error {
-	var newSampler Sampler
 	if probabilistic := res.GetProbabilisticSampling(); probabilistic != nil {
-		newSampler = newProbabilisticSampler(probabilistic.SamplingRate)
+		ps := newProbabilisticSampler(probabilistic.SamplingRate)
+		if !ps.Equal(s.sampler) {
+			s.sampler = ps
+		}
 	} else if rateLimiting := res.GetRateLimitingSampling(); rateLimiting != nil {
-		newSampler = NewRateLimitingSampler(float64(rateLimiting.MaxTracesPerSecond))
+		rls := NewRateLimitingSampler(float64(rateLimiting.MaxTracesPerSecond))
+		if !rls.Equal(s.sampler) {
+			s.sampler = rls
```

Comment:
we've already run into issues with this approach. Rate limiting sampler should not be just replaced, instead it's better to call Update() method on it that can reset the rate parameter while preserving & prorating the accumulated balance.